FEA Bisecting K-Means (#20031)
scikit-learn/scikit-learn: doc/modules/clustering.rst

 * :ref:`sphx_glr_auto_examples_cluster_plot_agglomerative_clustering_metrics.py`

Bisecting K-Means
-----------------

.. _bisect_k_means:

The :class:`BisectKMeans` is an iterative variant of :class:`KMeans`, using
divisive hierarchical clustering.
Instead of creating all centroids at once - It obtains desired number of centroids
by picking one cluster at each iteration and splitting it into two new clusters,
until target number of clusters is reached.

With that approach, it is possible to produce either an un-nested (flat)
clustering or a hierarchical clustering.
That approach is also preferable to agglomerative clustering
if the number of clusters is small, compared to the number of data points.

Cluster to split is picked depending on parameter ``bisect_strategy``:

> - Cluster with largest amount of data points (with ``largest_cluster``)

Reviewer: This does not render as a list. You might need to add two whitespaces before the dash.

- Cluster with biggest SSE (Sum of Squared Errors) (with ``biggest_sse``)

Picking by largest amount of data points in most cases produces result as
accurate as picking by SSE and is faster (especially for larger amount of data
points, where calculating error may be costly).

Difference between Bisect K-Means and regular K-Means can be seen on example below -
while regular K-Means tend to create non-related clusters, clusters from Bisect K-Means

> create quite visible hierarchy.

Reviewer: I don't understand where the visible hierarchy is. The example must be either more compelling or better explained. Adding some plot might be more intuitive, maybe the one in https://github.com/scikit-learn/scikit-learn/pull/20031#issuecomment-850650751 (see code in Details).

If the example has some plots or is more than a few lines, I would move the example in its own file `example/cluster/plot_bisect_k_means.py` (see other examples for expected syntax). Then I would link the rendered example to this guide (see other linked examples for the required syntax).

Author: @TomDLT I've added `examples/cluster/plot_kmeans_comparison.py`, but have problems with properly linking it. May I ask you for some tips?

Reviewer: You can get the link syntax from other examples in the page.

Direct link: 
```rst
:ref:`sphx_glr_auto_examples_cluster_plot_kmeans_comparison.py`
```
Link with a picture generated by the example:
```rst
.. image:: ../auto_examples/cluster/images/sphx_glr_plot_kmeans_comparison_001.png
    :target: ../auto_examples/cluster/plot_kmeans_comparison.html
    :scale: 38
```

Reviewer: However, I would rename the file to mention bisect kmeans, e.g. `plot_bisect_kmeans.py`.

Author: ah, ok - I did it in wrong way

 * :ref:`sphx_glr_auto_examples_cluster_plot_agglomerative_clustering_metrics.py`

Bisecting K-Means
-----------------

.. _bisect_k_means:

The :class:`BisectingKMeans` is an iterative variant of :class:`KMeans`, using
divisive hierarchical clustering.
Instead of creating all centroids at once, centroids are picked progressively based
on a previous clustering: a cluster is split into two new clusters repeatedly 
until the target number of clusters is reached.

This variant is preferable to agglomerative clustering
if the number of clusters is small compared to the number of data points.

This variant also do not produce empty clusters.

Cluster to split is picked depending on parameter ``bisect_strategy``:
 - Cluster with largest amount of data points (with ``largest_cluster``)

>  - Cluster with biggest SSE (Sum of Squared Errors) (with ``biggest_sse``)

Author: I've planned to change it a bit - now should be good

 * :ref:`sphx_glr_auto_examples_cluster_plot_agglomerative_clustering_metrics.py`

Bisecting K-Means
-----------------

.. _bisect_k_means:

The :class:`BisectingKMeans` is an iterative variant of :class:`KMeans`, using
divisive hierarchical clustering.
Instead of creating all centroids at once, centroids are picked progressively based
on a previous clustering: a cluster is split into two new clusters repeatedly 
until the target number of clusters is reached.

This variant is preferable to agglomerative clustering
if the number of clusters is small compared to the number of data points.

This variant also do not produce empty clusters.

There exist two strategies for selecting the cluster to split:

>  - ``bisect_strategy="largest_cluster"`` selects the cluster having the most points
>  - ``bisect_strategy="biggest_inerita"`` selects the cluster with biggest inertia

Reviewer 2: ```suggestion
 - ``bisecting_strategy="largest_cluster"`` selects the cluster having the most points
 - ``bisecting_strategy="biggest_inerita"`` selects the cluster with biggest inertia
```

>  - ``bisect_strategy="biggest_inerita"`` selects the cluster with biggest inertia

Reviewer: ```suggestion
 - ``bisecting_strategy="biggest_inertia"`` selects the cluster with biggest inertia
```

   (cluster with biggest Sum of Squared Errors within)

Picking by largest amount of data points in most cases produces result as

> accurate as picking by SSE and is faster (especially for larger amount of data

Reviewer: ```suggestion
accurate as picking by inertia and is faster (especially for larger amount of data
```

 * L. Hubert and P. Arabie, Comparing Partitions, Journal of
   Classification 1985
   <https://link.springer.com/article/10.1007%2FBF01908075>_

.. _bisect_k_means:

> Bisecting K-Means

Author: Not 100% sure if in correct place, but moved

=================

The :class:`BisectKMeans` is an iterative variant of :class:`KMeans`, using
divisive hierarchical clustering.
Instead of creating all centroids at once - It consequently split selected
cluster into two new clusters till it reaches desired number of clusters.

> With that approach, it is possible to produce either an un-nested (flat)

Author: Rewritten. Now should be more clear

clustering or a hierarchical clustering.

Cluster to split is picked depending on parameter ``bisect_strategy``:
- Cluster with largest amount of data points (with ``largest_cluster``)
- Cluster with biggest SSE (Sum of Squared Errors) (with ``biggest_sse``)

Cluster to split is always picked from all clusters obtained from previous

> iterations of algorithm. This leads to more balanced split of data.

Author: Unnecessary added. Removed with last commit
